invoices/sql_migration.go

```diff
@@ -366,3 +381,170 @@ func OverrideInvoiceTimeZone(invoice *Invoice) {
 		}
 	}
 }
+
+// MigrateInvoicesToSQL runs the migration of all invoices from the KV database
+// to the SQL database. The migration is done in a single transaction to ensure
+// that all invoices are migrated or none at all. This function can be run
+// multiple times without causing any issues as it will check if the migration
+// has already been performed.
+func MigrateInvoicesToSQL(ctx context.Context, db kvdb.Backend,
+	kvStore InvoiceDB, sqlStore *SQLStore, batchSize int) error {
+
+	offset := uint64(0)
+	var ops SQLInvoiceQueriesTxOptions
+	return sqlStore.db.ExecTx(ctx, &ops, func(tx SQLInvoiceQueries) error {
+		// First make sure that we haven't already migrated the
+		// invoices.
+		migration, err := tx.GetMigration(ctx, migrationID)
+		if err != nil {
+			return fmt.Errorf("unable to fetch migration '%v': %w",
+				migrationID, err)
+		}
+
+		// If there's a valid migration timestamp, then we've already
+		// migrated the invoices.
+		if migration.MigrationTime.Valid {
+			log.Infof("Migration '%v' has already been "+
+				"executed at %v", migrationID,
+				migration.MigrationTime.Time)
+
+			return nil
+		}
+
+		// Now create the hash index which we will use to look up
+		// invoice payment hashes by their add index during migration.
+		err = createInvoiceHashIndex(ctx, db, tx)
+		if err != nil && !errors.Is(err, ErrNoInvoicesCreated) {
+			log.Errorf("Unable to create invoice hash index: %v",
+				err)
+
+			return err
+		}
+
+		// Now we can start migrating the invoices. We'll do this in
+		// batches to reduce memory usage.
+		for {
+			query := InvoiceQuery{
+				IndexOffset:    offset,
+				NumMaxInvoices: uint64(batchSize),
+			}
+
+			queryResult, err := kvStore.QueryInvoices(ctx, query)
+			if err != nil && !errors.Is(err, ErrNoInvoicesCreated) {
+				return fmt.Errorf("unable to query invoices: "+
+					"%w", err)
+			}
+
+			if len(queryResult.Invoices) == 0 {
+				log.Infof("All invoices migrated")
+
+				break
+			}
+
+			err = migrateInvoices(
+				ctx, tx, sqlStore, queryResult.Invoices,
+			)
+			if err != nil {
+				return err
+			}
+
+			offset = queryResult.LastIndexOffset
+		}
+
+		// Clean up the hash index as it's no longer needed.
+		err = tx.ClearInvoiceHashIndex(ctx)
+		if err != nil {
+			return fmt.Errorf("unable to clear invoice hash "+
+				"index: %w", err)
+		}
```

Comment:
if we can get the sqlc/code-level migration ordering sorted, then this does not need to happen here, we can instead just add a new `up` file that drops the table & then we just need to mark that as the DB version _after_ this migration.

Reply:
Unfortunately we can't drop the table right now as there are queries depending on it and the query files are not versioned like the migration files. I think it's ok to keep the table there in an empty state until we can remove it in a later version though.

Reply:
> we can't drop the table right now as there are queries depending on it and the query files are not versioned like the migration files.

but we can technically have copied query files per migration like we do today for channeldb migrations yeah? ie, introduce some duplication in order to keep the live version of the interface clean?